src/autopas/containers/LeavingParticleCollector.h

```diff
@@ -9,38 +9,76 @@
 
 #include "autopas/options/IteratorBehavior.h"
 #include "autopas/particles/OwnershipState.h"
+#include "autopas/utils/ArrayMath.h"
 #include "autopas/utils/inBox.h"
 
-namespace autopas {
 /**
  * Namespace to collect leaving particles from a container.
  */
-namespace LeavingParticleCollector {
+namespace autopas::LeavingParticleCollector {
 /**
  * Collects leaving particles and marks halo particles as dummy.
  * @note This function does not move or actually delete any particles!
  * @tparam ContainerType The type of the container.
  * @param container The container from which particles should be collected.
  * @return Returns a vector of leaving particles.
- * @todo: openmp parallelization, general optimizations (do not iterate over all particles, but only over boundary,
- * eventually merge both loops?
  */
 template <class ContainerType>
 std::vector<typename ContainerType::ParticleType> collectParticlesAndMarkNonOwnedAsDummy(ContainerType &container) {
-  // First mark halo particles as dummy!
-  for (auto iter = container.begin(autopas::IteratorBehavior::halo); iter.isValid(); ++iter) {
-    iter->setOwnershipState(OwnershipState::dummy);
-  }
+  using namespace autopas::utils::ArrayMath::literals;
+
+  const auto upperBoundForMisplacement = container.getCutoff();
+  const auto containerDimensions = container.getBoxMax() - container.getBoxMin();
+  const auto lowerHaloCorner = container.getBoxMin() - upperBoundForMisplacement;
+  const auto upperHaloCorner = container.getBoxMax() + upperBoundForMisplacement;
+  // We need to extend these boundaries into the container because we need to find particles that are stored in the
+  // container but have since moved out. Extending by skin is not enough since we are interested in particles that have
```

Comment:
Regarding your conversation, the first comment is still a bit confusing. What I understand is this:

We extend the boundaries *outside* the container to find particles that have moved out and we extend the boundaries *into* the container to catch halo particles that moved in.

Reply:
Ah, now I get what I mean. Yeah, this is clearer